src/libostree/ostree-repo-commit.c

```diff
@@ -1867,7 +2033,8 @@ ostree_repo_write_content (OstreeRepo       *self,
                                           OSTREE_OBJECT_TYPE_FILE, &have_obj,
                                           cancellable, error))
         return FALSE;
-      if (have_obj)
+      if (have_obj &&
+          !(ostree_repo_get_mode (self) == OSTREE_REPO_MODE_ARCHIVE_Z2 && self->generate_sizes))
```

Comment:
This seems like the right thing to do, but actually a change from the old behavior. In the current single `write_object()` case, it seems that [having the correct object](https://github.com/endlessm/ostree/blob/da533f23c58fa8ed694456952ecef5d09bf95d37/src/libostree/ostree-repo-commit.c#L789) always returns early before `repo_store_size_entry`.

Actually, now that I look at it more, it seems like `ostree_repo_write_content` and `ostree_repo_write_metadata` are always called with a `NULL` expected checksum when creating commits, so you'd never reach here. On the other hand, checking `self->generate_sizes` seems like a better safeguard. In which case, I think this same condition should be added to `ostree_repo_write_metadata`.